modules/pricer/src/main/java/com/opengamma/strata/pricer/credit/cds/IsdaCdsProductPricer.java

```diff
@@ -0,0 +1,743 @@
+      ResolvedCds cds,
+      CreditRatesProvider ratesProvider,
+      LocalDate referenceDate,
+      PriceType priceType,
+      ReferenceData refData) {
+
+    double riskyAnnuity = riskyAnnuity(cds, ratesProvider, referenceDate, priceType, refData);
+    return CurrencyAmount.of(cds.getCurrency(), cds.getBuySell().normalize(cds.getNotional()) * riskyAnnuity);
+  }
+
+  /**
+   * Calculates the recovery01 of the CDS product.
+   * <p>
+   * The recovery01 is defined as the present value sensitivity to the recovery rate.
+   * Since the ISDA standard model requires the recovery rate to be constant throughout the lifetime of the CDS,  
+   * one currency amount is returned by this method.
+   * 
+   * @param cds  the product
+   * @param ratesProvider  the rates provider
+   * @param referenceDate  the reference date
+   * @param refData  the reference data
+   * @return the recovery01
+   */
+  public CurrencyAmount recovery01(
+      ResolvedCds cds,
+      CreditRatesProvider ratesProvider,
+      LocalDate referenceDate,
+      ReferenceData refData) {
+
+    ArgChecker.notNull(cds, "cds");
+    ArgChecker.notNull(ratesProvider, "ratesProvider");
+    ArgChecker.notNull(referenceDate, "referenceDate");
+    ArgChecker.notNull(refData, "refData");
+    if (!cds.getProtectionEndDate().isAfter(ratesProvider.getValuationDate())) { //short cut already expired CDSs
+      return CurrencyAmount.of(cds.getCurrency(), 0d);
+    }
+    LocalDate stepinDate = cds.getStepinDateOffset().adjust(ratesProvider.getValuationDate(), refData);
+    LocalDate effectiveStartDate = cds.getEffectiveStartDate(stepinDate);
+    recoveryRate(cds, ratesProvider); // for validation
+    Pair<CreditDiscountFactors, LegalEntitySurvivalProbabilities> rates = reduceDiscountFactors(cds, ratesProvider);
+    double protectionFull =
+        protectionFull(cds, rates.getFirst(), rates.getSecond(), referenceDate, effectiveStartDate);
+
+    return CurrencyAmount.of(cds.getCurrency(), -cds.getBuySell().normalize(cds.getNotional()) * protectionFull);
+  }
+
+  /**
+   * Calculates the present value sensitivity of the product. 
+   * <p>
+   * The present value sensitivity of the product is the sensitivity of present value to the underlying curves.
+   * 
+   * @param cds  the product 
+   * @param ratesProvider  the rates provider
+   * @param referenceDate  the reference date
+   * @param refData  the reference data
+   * @return the present value sensitivity
+   */
+  public PointSensitivityBuilder presentValueSensitivity(
+      ResolvedCds cds,
+      CreditRatesProvider ratesProvider,
+      LocalDate referenceDate,
+      ReferenceData refData) {
+
+    ArgChecker.notNull(cds, "cds");
+    ArgChecker.notNull(ratesProvider, "ratesProvider");
+    ArgChecker.notNull(referenceDate, "referenceDate");
+    ArgChecker.notNull(refData, "refData");
+    if (!cds.getProtectionEndDate().isAfter(ratesProvider.getValuationDate())) { //short cut already expired CDSs
```

Comment:
This code occurs lots of times. Create `isExpired(cds, ratesProvider)` and update all the usages, then there is no need for the comment.